mermaid_py/client.py

```diff
@@ -0,0 +1,443 @@
+        if token:
+            self.authenticated = True
+        # Initializes requests Session and assigns headers for all Client MERMAID API calls.
+        self.session = requests.Session()
+        self.session.headers.update(
+            {
+                "content-type": "application/json",
+                "authorization": "Bearer %s" % self.token,
+            }
+        )
+
+    # API paths
+    def _fetch_resource(self, resource=None, parameters=None, data=None):
+        """
+        Prepares API call and utilizes Client Session to make MERMAID API calls.
+        :param resource: (optional) subdirectory path. Defaults to (subdirectory='').
+        :type resource: str
+        :param parameters: (optional) parameters for request.
+        :type parameters: dict:(../?key=val), str:(../?str).
+        :param data: (optional) data. Defaults to (data=None).
+        :return: JSON object containing MERMAID API data.
+        :rtype: dict
+        """
+        # Creates full URL path.
+        if resource is None:
+            raise Exception(f"_fetch_resource invalid resource:{resource}")
+
+        prep_url = self.url + resource
+        # Prepares Request and send from Client class Session.
+        req = requests.Request("GET", url=prep_url, params=parameters, data=data)
+        prepped = self.session.prepare_request(req)
+        resp = self.session.send(prepped)
+
+        # Returns JSON if response code OK.
+        if resp.status_code == requests.codes.ok:
+            return resp.json()
+        else:
+            if resp.status_code == 401:
+                raise Exception(
+                    f"_fetch_resource response code: {resp.status_code} -- Attempt token refresh"
+                )
+
+            else:
+                raise Exception(f"_fetch_resource response code: {resp.status_code}")
+
+    def _api_projects_path(
+        self, name=None, id=None, resource=None, filter=None, filter_val=None
+    ):
+        """
+        Helper class function used to create and call projects path '/projects/<project_id>/' based on name or id
+        of project. See https://mermaid-api.readthedocs.io/en/latest/projects.html#project-resources for more
+        information.
+        :param name: (optional if 'id' provided) MERMAID project name.
+        :type name: str
+        :param id: (optional if 'name' provided) MERMAID project ID.
+        :type: id: str
+        :param resource: MERMAID project entity resources. /projects/<project_id>/<resource>
+        :param filter: MERMAID project resource filters.
+        :type filter: str
+        :param filter_val: (optional) Required for resource filters requiring values. eg. (size_min/size_max,
+        count_min/count_max, length_min/length_max, len_surveyed_min/len_surveyed_max,
+        sample_date_before/sample_date_after).
+        :type filter_val: str, int
+        :return: project resource.
+        :rtype: dict
+        """
+        # Generate path.
+        path = None
```

Comment:
See https://github.com/data-mermaid/mermaid-py/pull/1#discussion_r686246386

Reply:
Kept your logic but made 'id' favored to 'name' by only calling name if 'id' is None to prevent an extra API call.